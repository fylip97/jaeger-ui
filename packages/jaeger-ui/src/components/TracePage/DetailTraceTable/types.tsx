export type TableSpan = {
  name: string;
  count: number;
  total: number;
  avg: number;
  min: number;
  max: number;
  isDetail: boolean;
  self: number;
  selfAvg: number;
  selfMin: number;
  selfMax: number;
  percent: number;
  key: string;
  child: boolean;
  parentElement: string;
  color: string;
  searchColor: string;

}
