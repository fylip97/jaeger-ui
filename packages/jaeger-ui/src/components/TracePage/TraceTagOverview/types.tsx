export type TableSpan = {
  name: string;
  count: number;
  total: number;
  avg: number;
  min: number;
  max: number;
  exc: number;
  excAvg: number;
  excMin: number;
  excMax: number;
  percent: number;
  isDetail: boolean;
  parentElement: string;

}