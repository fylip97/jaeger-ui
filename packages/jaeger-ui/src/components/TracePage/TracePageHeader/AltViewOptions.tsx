// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button, Dropdown, Icon, Menu } from 'antd';
import { Link } from 'react-router-dom';

import {
  trackGanttView,
  trackGraphView,
  trackStatisticsView,
  trackJsonView,
  trackRawJsonView,
  trackAnalyseView,
} from './TracePageHeader.track';
import prefixUrl from '../../../utils/prefix-url';

type Props = {
  onTraceGraphViewClicked: (index: number) => void;
  traceID: string;
  selectedTraceView: number;
};

export default function AltViewOptions(props: Props) {

  const { onTraceGraphViewClicked, selectedTraceView, traceID } = props;
  const menuItems = ['Trace Timeline', 'Trace Graph', 'Trace Analyse', 'Trace Statistics'];

  const handleToggleView = (index: number, item: string) => {
    if (item === menuItems[0]) {
      trackGanttView();
    } else if (item === menuItems[1]) {
      trackGraphView();
    } else if (item === menuItems[2]) {
      trackAnalyseView()
    } else if (item === menuItems[3]) {
      trackStatisticsView();
      
    }
    onTraceGraphViewClicked(index);
  };

  const toggle = () => {
    let nextIndex = selectedTraceView + 1;
    if (nextIndex > 3) {
      nextIndex = 0;
    }
    if (nextIndex === 0) {
      trackGanttView();
    } else if (nextIndex === 1) {
      trackGraphView();
    } else if (nextIndex === 2) {
      trackAnalyseView();
    } else if (nextIndex === 3) {
      trackStatisticsView();
    }
    onTraceGraphViewClicked(nextIndex);
  };

  const menu = (
    <Menu>
      {menuItems.map((item, index) =>
        index === selectedTraceView ? null : (
          <Menu.Item key={item}>
            <a onClick={() => handleToggleView(index, item)} role="button">
              {item}
            </a>
          </Menu.Item>
        )
      )}
      <Menu.Item>
        <Link
          to={prefixUrl(`/api/traces/${traceID}?prettyPrint=true`)}
          rel="noopener noreferrer"
          target="_blank"
          onClick={trackJsonView}
        >
          Trace JSON
        </Link>
      </Menu.Item>
      <Menu.Item>
        <Link
          to={prefixUrl(`/api/traces/${traceID}?raw=true&prettyPrint=true`)}
          rel="noopener noreferrer"
          target="_blank"
          onClick={trackRawJsonView}
        >
          Trace JSON (unadjusted)
        </Link>
      </Menu.Item>
    </Menu>
  );

  return (
    <Dropdown overlay={menu}>
      <Button onClick={toggle} className="ub-mr2" htmlType="button">
        Alternate Views <Icon type="down" />
      </Button>
    </Dropdown>
  );
}
