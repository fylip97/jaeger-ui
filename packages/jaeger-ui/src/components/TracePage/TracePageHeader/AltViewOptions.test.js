// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import { Dropdown } from 'antd';
import { Link } from 'react-router-dom';
import AltViewOptions from './AltViewOptions';
import * as track from './TracePageHeader.track';
import { ETraceViewType } from '../types';

describe('AltViewOptions', () => {
  let trackGanttView;
  let trackGraphView;
  let trackJsonView;
  let trackRawJsonView;
  let trackStatisticsView;

  let wrapper;
  const getLink = text => {
    const menu = shallow(wrapper.find(Dropdown).prop('overlay'));
    const links = menu.find(Link);
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i);
      if (link.children().text() === text) return link;
    }
    const links2 = menu.find('a');
    for (let i = 0; i < links.length; i++) {
      const link = links2.at(i);
      if (link.children().text() === text) return link;
    }
    throw new Error(`Could not find "${text}"`);
  };

  const props = {
    viewType: ETraceViewType.TraceTimelineViewer,
    traceID: 'test trace ID',
    onTraceViewChange: jest.fn(),
  };

  beforeAll(() => {
    trackGanttView = jest.spyOn(track, 'trackGanttView');
    trackGraphView = jest.spyOn(track, 'trackGraphView');
    trackJsonView = jest.spyOn(track, 'trackJsonView');
    trackRawJsonView = jest.spyOn(track, 'trackRawJsonView');
    trackStatisticsView = jest.spyOn(track, 'trackStatisticsView');
  });

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AltViewOptions {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('tracks viewing JSONs', () => {
    expect(trackJsonView).not.toHaveBeenCalled();
    getLink('Trace JSON').simulate('click');
    expect(trackJsonView).toHaveBeenCalledTimes(1);

    expect(trackRawJsonView).not.toHaveBeenCalled();
    getLink('Trace JSON (unadjusted)').simulate('click');
    expect(trackRawJsonView).toHaveBeenCalledTimes(1);

    expect(trackJsonView).toHaveBeenCalledTimes(1);
    expect(trackGanttView).not.toHaveBeenCalled();
    expect(trackGraphView).not.toHaveBeenCalled();
  });

  it('track dropdown menu', () => {
    expect(trackGraphView).not.toHaveBeenCalled();
    expect(props.onTraceViewChange).not.toHaveBeenCalled();
    getLink('Trace Graph').simulate('click');
    expect(trackGraphView).toHaveBeenCalledTimes(1);
    expect(props.onTraceViewChange).toHaveBeenCalledTimes(1);
    expect(trackStatisticsView).not.toHaveBeenCalled();
    getLink('Trace Statistics').simulate('click');
    expect(trackStatisticsView).toHaveBeenCalledTimes(1);
    expect(props.onTraceViewChange).toHaveBeenCalledTimes(2);

    wrapper.setProps({ viewType: ETraceViewType.TraceGraph });
    expect(trackGanttView).not.toHaveBeenCalled();
    getLink('Trace Timeline').simulate('click');
    expect(trackGanttView).toHaveBeenCalledTimes(1);
    expect(props.onTraceViewChange).toHaveBeenCalledTimes(3);
    getLink('Trace Statistics').simulate('click');
    expect(trackStatisticsView).toHaveBeenCalledTimes(2);
    expect(props.onTraceViewChange).toHaveBeenCalledTimes(4);

    wrapper.setProps({ viewType: 2 });
    getLink('Trace Timeline').simulate('click');
    expect(trackGanttView).toHaveBeenCalledTimes(2);
    expect(props.onTraceViewChange).toHaveBeenCalledTimes(5);
    getLink('Trace Graph').simulate('click');
    expect(trackGraphView).toHaveBeenCalledTimes(2);
    expect(props.onTraceViewChange).toHaveBeenCalledTimes(6);
  });
});
