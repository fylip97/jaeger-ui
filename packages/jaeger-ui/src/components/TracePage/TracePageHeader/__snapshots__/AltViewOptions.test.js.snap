// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AltViewOptions renders correctly 1`] = `
<Dropdown
  mouseEnterDelay={0.15}
  mouseLeaveDelay={0.1}
  overlay={
    <Menu
      className=""
      focusable={false}
      prefixCls="ant-menu"
      theme="light"
    >
      <MenuItem>
        <a
          onClick={[Function]}
          role="button"
        >
          Trace Graph
        </a>
      </MenuItem>
      <MenuItem>
        <a
          onClick={[Function]}
          role="button"
        >
          Trace Statistics
        </a>
      </MenuItem>
      <MenuItem>
        <Link
          onClick={[MockFunction]}
          rel="noopener noreferrer"
          replace={false}
          target="_blank"
          to="/api/traces/test trace ID?prettyPrint=true"
        >
          Trace JSON
        </Link>
      </MenuItem>
      <MenuItem>
        <Link
          onClick={[MockFunction]}
          rel="noopener noreferrer"
          replace={false}
          target="_blank"
          to="/api/traces/test trace ID?raw=true&prettyPrint=true"
        >
          Trace JSON (unadjusted)
        </Link>
      </MenuItem>
    </Menu>
  }
  placement="bottomLeft"
  prefixCls="ant-dropdown"
>
  <Button
    block={false}
    className="ub-mr2"
    ghost={false}
    htmlType="button"
    loading={false}
    prefixCls="ant-btn"
  >
    Trace Timeline
    <Icon
      type="down"
    />
  </Button>
</Dropdown>
`;
